devito/ir/iet/visitors.py

```diff
@@ -372,22 +373,74 @@ def visit_Call(self, o, ret=None, queue=None):
     visit_Element = FindSections.visit_Node
 
 
-class MapIteration(FindSections):
+class MapNodes(Visitor):
+
+    @classmethod
+    def default_retval(cls):
+        return OrderedDict()
 
     """
-    Map each :class:`Iteration` object in the Iteration/Expression tree to the
-    enclosed :class:`Expression` and :class:`Call` objects.
+    Given an Iteration/Expression tree, build a mapper between parent and
+    children nodes of given type.
+
+    Parameters
+    ----------
+    parent_type : Node, optional
+        By default, parents are of type Iteration. One can alternatively supply
+        a different type.
+    child_types : Node or list of Node, optional
+        By default, children of type Call and Expression are retrieved.
+        One can alternatively supply one or more different types.
+    mode : str, optional
+        By default, a depth-first visit is performed. Alternatively, if the
+        keyword 'immediate' is passed, the visit stops to the immediate children
+        of a given parent.
+    groupby : bool, optional
+        If True, nested nodes of type ``parent_type`` are used as key in the
+        built mapper. Defaults to False.
     """
 
-    def visit_Call(self, o, ret=None, queue=None):
+    def __init__(self, parent_type=None, child_types=None, mode=None, groupby=False):
+        super(MapNodes, self).__init__()
+        self.parent_type = parent_type or Iteration
+        self.child_types = as_tuple(child_types) or (Call, Expression)
+        assert mode is None or mode == 'immediate'
+        self.mode = mode
+        self.groupby = groupby
```

Comment:
private `_` and properties?

Reply:
there is no public and no private in the IET visitors. We could add them though...